File: hw/vargas/u3/hw29Singleton/javaScript/ec/edu/espe/singleton/utils/GmailConfiguration.js
import { MongoManager } from './MongoManager.js';

export class GmailConfiguration {
    static #instance = null;

    constructor() {
        if (GmailConfiguration.#instance) {
            throw new Error("Use GmailConfiguration.getInstance()");
        }
        this.senderEmail = null;
        this.appPassword = null;
        this.smtpHost = "smtp.gmail.com";
        this.smtpPort = 465;
        this.configLoaded = false;
    }

    static getInstance() {
        if (!GmailConfiguration.#instance) {
            GmailConfiguration.#instance = new GmailConfiguration();
        }
        return GmailConfiguration.#instance;
    }

    async loadConfigurationFromDatabase() {
        if (this.configLoaded) return;

        const mongoManager = new MongoManager();
        const results = await mongoManager.find("system_config", {});

        if (results && results.length > 0) {
            const configDoc = results[0];
            this.senderEmail = configDoc.sender_email;
            this.appPassword = configDoc.app_password;
            this.smtpHost = configDoc.smtp_host || "smtp.gmail.com";
            this.smtpPort = parseInt(configDoc.smtp_port) || 465;
            this.configLoaded = true;
        } else {
            console.error("Advertencia: No se encontró configuración de email en la base de datos.");
        }
        
        await mongoManager.close();
    }
}

File: hw/vargas/u3/hw29Singleton/python/ec/edu/espe/singleton/utils/gmail_configuration.py
from .mongo_manager import MongoManager

class GmailConfiguration:

    _instance = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super(GmailConfiguration, cls).__new__(cls)
            cls._instance._load_configuration_from_database()
        return cls._instance

    def _load_configuration_from_database(self):
        self.sender_email = None
        self.app_password = None
        self.smtp_host = "smtp.gmail.com"
        self.smtp_port = 465
        
        mongo_manager = MongoManager()
        results = mongo_manager.find("system_config", {})

        if results:
            config_doc = results[0]
            self.sender_email = config_doc.get("sender_email")
            self.app_password = config_doc.get("app_password")
            self.smtp_host = config_doc.get("smtp_host", "smtp.gmail.com")
            self.smtp_port = int(config_doc.get("smtp_port", 465))
        else:
            print("Advertencia: No se encontró configuración de email en la base de datos.")

    @classmethod
    def get_instance(cls):
        return cls()